adoptingerlang.org

```diff
@@ -1331,6 +1331,364 @@ Repeat this exercise from time to time, and eventually do some [[https://princip
 :EXPORT_HUGO_MENU: :menu main :parent development
 :END:
 
+One major distinction between programming 20 years ago and programming today is in how software projects shifted in their approaches. Whereas material of yesterday focused a lot on writing reusable and extendable components within each project, the current trend is to make smaller, isolated projects that each individually ends up being easy to throw away and replace. Current microservice implementations and Javascript dependency trees may lead us to believe no project is too small to be discardable.
+
+Ironically, this approach towards strong isolation and clearly-defined boundaries—whether they be interfaces, network APIs, or protocols—that was necessary for small single-purpose components has given rise to what is possibly the greatest amount of code reuse we've ever seen through libraries and packages. Reuse materialized from shared needs across various people and projects, not from the mythical ability of providing the perfectly extendable class hierarchy. There is so much reuse today that we have reached the point where folks are starting to ask if maybe, just maybe, we aren't reusing more code than we should: each library comes with risks and liabilities, and we're exposing ourselves to a lot of it for the sake of moving fast.
+
+Erlang, for its own part, has always lived and breathed strong isolation with clearly defined message-passing protocols (often wrapped behind functional interfaces). Its small community size of generally experienced developers probably made it lag behind other communities when it comes to libraries and packaging. It was not until 2009 that it became easy to install libraries from other people, through the first version of =rebar=. Erlang libraries proliferated all over GitHub and similar pages haphazardly, and at some point, it was possible to find over twelve different versions of the same PostgreSQL driver, which all had the same name and similar versions, but all did a few things differently. Multiple people tried to write package managers that would simplify the installation of dependencies, but none managed take hold.
```

Comment:
I think a large part of what made rebar catch on (resulting in it being easy to install libraries from other people) was being from Basho, that and being an escript. So I'd phrase it as "Despite the efforts of CEAN from ProcessOne and faxien, no package manager gained traction in the community, it was not until 2009 that it really became easy to install libraries from other people, through the first version of rebar."

Reply:
Oh, maybe also a good thing to mention in these first few paragraphs and that rebar brought new to the ecosystem, is the move to per-project dependency installation. That is something that changed across the industry and pretty sure rebar was the first to introduce it to Erlang, while faxien and CEAN installed applications into the OTP lib.